chore(email): Switch to \Zend\Mail library instead of raw mail() function

ewinslow opened the pull request:
This lays a good foundation for improving Elgg's flexibility and security
in sending emails. Some of the features we open up because of this change:
- HTML emails
- Attachments
- Improved security (automatic header escaping, etc.)
- Third-party SMTP services (e.g. Mandrill)
- Receiving emails
- In-memory transport for easier testing/development

We aren't introducing our own interfaces because `Zend\Mail`
is plenty well supported and the API is agreeable to us.
Writing a wrapper would just be unnecessary overhead.

BREAKING CHANGE:
We are switching to `Zend\Mail` for sending emails in Elgg 2.0.
It's likely that there are some edge cases that the library
handles differently than Elgg 1.x used to. Take care to test
your email notifications carefully when upgrading to 2.0.

Fixes #5918

ewinslow committed:
chore(email): Switch to `Zend\Mail` instead of built-in `mail()` function

This lays a good foundation for improving Elgg's flexibility and security
in sending emails. Some of the features we open up because of this change:

 * HTML emails
 * Attachments
 * Improved security (automatic header escaping, etc.)
 * Third-party SMTP services (e.g. Mandrill)
 * Receiving emails
 * In-memory transport for easier testing/development

We aren't introducing our own interfaces because `Zend\Mail`
is plenty well supported and the API is agreeable to us.
Writing a wrapper would just be unnecessary overhead.

BREAKING CHANGE:
We are switching to `Zend\Mail` for sending emails in Elgg 2.0.
It's likely that there are some edge cases that the library
handles differently than Elgg 1.x used to. Take care to test
your email notifications carefully when upgrading to 2.0.

Fixes #5918

ewinslow commented:
Note: this is UNTESTED atm.

ewinslow commented:
When we announce this we can focus on the benefits of:
 * HTML email support
 * Attachment support
 * Improved security (automatic header escaping, etc.)
 * Performance
   * Can utilize a third party transport like Mandrill instead of localhost
   * Can load-balance transports

mrclay commented on engine/lib/notification.php:
```diff
@@ -623,62 +642,25 @@ function elgg_send_email($from, $to, $subject, $body, array $params = null) {
-	}
-
-	// Format message
-	$body = html_entity_decode($body, ENT_QUOTES, 'UTF-8'); // Decode any html entities
```
Unless we're switching all to html mail, I think we need these lines.

mrclay replied:
Particularly HTML decode followed by stripping tags. Subject needs the same.

mrclay commented:
Strawman based on my "env" service PR: in settings.php allow `$CONFIG->mailTransportFactory` to be a Closure that returns a valid transport.

mrclay commented:
Actually I wonder if we need to reserve `$CONFIG->factories` as an array of Closures for injecting services (but we only handle those we support, not arbitrary keys).

ewinslow commented:
Originally I had considered zend's library. I'm still open to that. The attachments api isn't super nice but they have support for receiving mail which seems killer.

mrclay commented on engine/lib/notification.php:
```diff
@@ -577,6 +577,35 @@ function set_user_notification_setting($user_guid, $method, $value) {
+	$containsName = preg_match('/<(.*)>/', $contact, $matches) == 1;
+	if ($containsName) {
+		return [
+			'name' => trim(elgg_substr($contact, 0, $email_pos)),
```
Where does $email_pos come from? If $containsName, just use the same pattern with preg_replace to remove it and trim the result?

ewinslow replied:
Should be `elgg_strpos($contact, '<')`

mrclay commented on engine/lib/notification.php:
```diff
@@ -593,16 +622,6 @@ function set_user_notification_setting($user_guid, $method, $value) {
 function elgg_send_email($from, $to, $subject, $body, array $params = null) {
 	global $CONFIG;
 
-	if (!$from) {
```
Why are these checks going away?

ewinslow replied:
Over eager cleanup

ewinslow replied:
Will restore

mrclay commented:
Obviously needs mention in docs/guides/upgrading

beck24 commented:
The switch to SwiftMailer is good, should we also default to html emails?  We're at a point where every email client handles them just fine and *every* project I work on requires html emails as a feature.

ewinslow commented:
Switching to HTML emails sounds great and this will make that possible without hacks, but I think that can/should come as a separate effort.

mrclay commented:
...As in get this well integrated, then attack HTML in a separate PR. Sounds good.

ewinslow commented:
More reasons I've reverted to my preference for Zend's Mail library, in addition to the ability to read/receive emails.

 * In-memory transport which can be used during testing/development.
 * Provides a ridiculously simple mailer/transport interface: a single `send(Message $message)` method

ewinslow commented:
OK, I believe I addressed all questions with this update. Some changes I made:

 * Switched to Zend, since it offers more features and has a cleaner API, in my estimation.
 * Moved new private helper functions to static methods on a new `Elgg\Mail\Address` helper class.
 * Added a test for this function (it's actually possible now!)
 * Added docs to upgrading guide
 * Restored previously deleted functionality to elgg_send_email

mrclay commented on engine/lib/notification.php:
```diff
@@ -623,62 +628,33 @@ function elgg_send_email($from, $to, $subject, $body, array $params = null) {
 	$body = elgg_strip_tags($body); // Strip tags from message
-	$body = preg_replace("/(\r\n|\r)/", "\n", $body); // Convert to unix line endings in body
-	$body = preg_replace("/^From/", ">From", $body); // Change lines starting with From to >From
 	$body = wordwrap($body);
```
Does Zend not do this automatically?

mrclay commented on engine/tests/phpunit/Elgg/Mail/MailerTest.php:
```diff
@@ -0,0 +1,24 @@
+class MailerTest extends TestCase {
+	
+	function testElggSendEmailPassesAllFieldsAsMessageToMailer() {
+		$mailer = new InMemoryTransport();
```
yeah, this is awesome

mrclay replied:
We should inject this in the PHPUnit bootstrap

mrclay commented on engine/lib/notification.php:
```diff
@@ -623,62 +628,33 @@ function elgg_send_email($from, $to, $subject, $body, array $params = null) {
-	return mail($to, $subject, $body, $headers_string);
+	
+	$message = new Message();
+	$message->addFrom($from_address);
```
need to set encoding

ewinslow commented:
@mrclay Can you submit a PR to my PR, or perhaps just LGTM this and follow up with another one later to fix these bugs? They both seem like good ideas but not blockers, and I'm a little busy these days with the :baby:

mrclay commented:
I sent you https://github.com/ewinslow/Elgg/pull/6

juho-jaakkola commented:
I don't have time to review right now, but I'm definitely not blocking.

mrclay commented:
I'm going to make a new PR with this + https://github.com/ewinslow/Elgg/pull/6.

mrclay commented:
Replaced by #8331
